spyder/plugins/application/widgets/install.py

```diff
@@ -0,0 +1,258 @@
+            self.accept()
+
+    def close_installer(self):
+        """Close the installation dialog."""
+        if (self.status == FINISHED
+                or self.status == CANCELLED):
+            self.setup()
+            self.accept()
+        else:
+            self.hide()
+
+    def reject(self):
+        """Reimplement Qt method."""
+        on_installation_widget = self._installation_widget.isVisible()
+        if on_installation_widget:
+            self.close_installer()
+        else:
+            super(UpdateInstallerDialog, self).reject()
+
+    def _change_update_installation_status(self, status=NO_STATUS):
+        """Set the installation status."""
+        self.status = status
+        self.sig_installation_status.emit(self.status)
+
+    def _progress_reporter(self, block_number, read_size, total_size):
+        progress = 0
+        if total_size > 0:
+            progress = block_number * read_size
+        if self.cancelled:
+            raise UpdateInstallationCancelledException()
+        else:
+            self.sig_download_progress.emit(progress, total_size)
+
+    def cancell_thread_install_update(self):
+        self._change_update_installation_status(status=CANCELLED)
+        self.thread_install_update.join()
+
+    def _download_install(self):
+        try:
+            with TemporaryDirectory(prefix="Spyder-") as tmpdir:
+                destination = os.path.join(tmpdir, 'updateSpyder.exe')
+                url = (
+                    'https://github.com/spyder-ide/spyder/releases/latest/'
+                    'download/Spyder_64bit_lite.exe')
+                val = (is_module_installed('numpy') or
+                       is_module_installed('pandas'))
+                if val:
+                    url = (
+                        'https://github.com/spyder-ide/spyder/releases/latest/'
+                        'download/Spyder_64bit_full.exe')
+                download = urlretrieve(url,
+                                       destination,
+                                       reporthook=self._progress_reporter)
+                self._change_update_installation_status(status=INSTALLING)
+                install = subprocess.Popen(destination, shell=True)
+                install.communicate()
+        except UpdateInstallationCancelledException:
+            self._change_update_installation_status(status=CANCELLED)
+        finally:
+            self._change_update_installation_status(status=PENDING)
+
+    def start_installation_update(self):
+        self.cancelled = False
+        self._change_update_installation_status(
+            status=DOWNLOADING_INSTALLER)
+        """call a function in a simple thread, to prevent blocking"""
+        self.thread_install_update = threading.Thread(
+            target=self._download_install)
+        self.thread_install_update.start()
+
+
+class UpdateInstallationCancelledException(Exception):
+    """Update installation was cancelled."""
+    pass
```

Comment:
Please move this class to be before `UpdateInstallation`. We tend define support classes before others and not after.

Reply:
This is still not addressed.